packages/firebase_auth/firebase_auth_web/lib/firebase_auth_web.dart

```diff
@@ -0,0 +1,352 @@
+import 'package:firebase/firebase.dart' as firebase;
+import 'package:flutter/services.dart' show PlatformException;
+import 'package:flutter_web_plugins/flutter_web_plugins.dart';
+
+class FirebaseAuthWeb extends FirebaseAuthPlatform {
+  static void registerWith(Registrar registrar) {
+    FirebaseAuthPlatform.instance = FirebaseAuthWeb();
+  }
+
+  firebase.Auth _getAuth(String name) {
+    final firebase.App app = firebase.app(name);
+    return firebase.auth(app);
+  }
+
+  PlatformAdditionalUserInfo _fromJsAdditionalUserInfo(
+      firebase.AdditionalUserInfo additionalUserInfo) {
+    return PlatformAdditionalUserInfo(
+      isNewUser: additionalUserInfo.isNewUser,
+      providerId: additionalUserInfo.providerId,
+      username: additionalUserInfo.username,
+      profile: additionalUserInfo.profile,
+    );
+  }
+
+  PlatformUserInfo _fromJsUserInfo(firebase.UserInfo userInfo) {
+    return PlatformUserInfo(
+      providerId: userInfo.providerId,
+      uid: userInfo.providerId,
+      displayName: userInfo.displayName,
+      photoUrl: userInfo.photoURL,
+      email: userInfo.email,
+      phoneNumber: userInfo.phoneNumber,
+    );
+  }
+
+  PlatformUser _fromJsUser(firebase.User user) {
+    return PlatformUser(
+      providerId: user.providerId,
+      uid: user.uid,
+      displayName: user.displayName,
+      photoUrl: user.photoURL,
+      email: user.email,
+      phoneNumber: user.phoneNumber,
+      creationTimestamp:
+          DateTime.parse(user.metadata.creationTime).millisecondsSinceEpoch,
+      lastSignInTimestamp:
+          DateTime.parse(user.metadata.lastSignInTime).millisecondsSinceEpoch,
+      isAnonymous: user.isAnonymous,
+      isEmailVerified: user.emailVerified,
+      providerData:
+          user.providerData.map<PlatformUserInfo>(_fromJsUserInfo).toList(),
+    );
+  }
+
+  PlatformAuthResult _fromJsUserCredential(firebase.UserCredential credential) {
+    return PlatformAuthResult(
+      user: _fromJsUser(credential.user),
+      additionalUserInfo: _fromJsAdditionalUserInfo(
+        credential.additionalUserInfo,
+      ),
+    );
+  }
+
+  PlatformIdTokenResult _fromJsIdTokenResult(
+      firebase.IdTokenResult idTokenResult) {
+    return PlatformIdTokenResult(
+      token: idTokenResult.token,
+      expirationTimestamp: idTokenResult.expirationTime.millisecondsSinceEpoch,
+      authTimestamp: idTokenResult.authTime.millisecondsSinceEpoch,
+      issuedAtTimestamp: idTokenResult.issuedAtTime.millisecondsSinceEpoch,
+      claims: idTokenResult.claims,
+      signInProvider: idTokenResult.signInProvider,
+    );
+  }
+
+  firebase.User _getCurrentUserOrThrow(firebase.Auth auth) {
+    final firebase.User user = auth.currentUser;
+    if (user == null) {
+      throw PlatformException(
+        code: 'USER_REQUIRED',
+        message: 'Please authenticate with Firebase first',
+      );
+    }
+    return user;
+  }
+
+  firebase.OAuthCredential _getCredential(AuthCredential credential) {
```

Comment:
We can export an app-facing `OAuthCredential` class and use it in the platform interface since that class exists in the iOS and Android SDKs.

https://firebase.google.com/docs/reference/android/com/google/firebase/auth/OAuthCredential

However, I think it should be a separate class from the platform implementer's view which is aware of the package:firebase version of the class.

I think the way that you get an `OAuthCredential` is casting the result of `getCredential` if you know you got it from a provider that is OAuth-based.

Reply:
My dart-fu is not very strong, but in order to be able to cast like that, you'd still need to "extend" the `firebase.OAuthCredential` class somehow, right? Or end up with a method like this one?

I think this is a good tradeoff between the platform_interface and package:firebase, as it is.

Reply:
I would prefer not to expose `OAuthCredential` at all. Unfortunately `AuthCredential` was removed from `package:firebase` in this PR: https://github.com/FirebaseExtended/firebase-dart/pull/255

I plan on adding `AuthCredential` back and making `OAuthCredential` a subclass, just like it is in the Firebase SDK

Reply:
@hterkelsen that approach SGTM. 

We should not expose the package:firebase version of `OAuthCredential` to app developers because we don't want package:firebase to be a dependency on mobile. However we should have an `AuthCredential` and `OAuthCredential` to match the iOS and Android SDKs, defined in the platform interface and exported to be app-facing.